### DISCUSSION TITLE
How to use `installShellFiles` with the new API ?

### DISCUSSION THREAD

**Discussion by pinage404** [90-008/nix-cargo-integration#104] [Q&A]
Hello,

I tried to upgrade Rust from 1.64 to 1.65 but it fails https://gitlab.com/pinage404/git-gamble/-/jobs/3342573954 ; i think it's because i use an old version of nix-cargo-integration

I'm trying to upgrade from https://github.com/yusdacra/nix-cargo-integration/tree/3df28172b283d56c5b4f08d0291087a12da7c6f1 (previous main branch) to https://github.com/yusdacra/nix-cargo-integration/tree/5cc9744efb05e9bd4f0760e487f177eb47b74dac (current main branch)

I don't understand how to upgrade

Example in https://yusdacra.github.io/nix-cargo-integration/pkg-config-options.html#overrides seems to not match the templates https://github.com/yusdacra/nix-cargo-integration/blob/master/templates/full/flake.nix (`overrideAttrs` vs `override`)

Tries :
https://gitlab.com/pinage404/git-gamble/-/jobs/3350297601
https://gitlab.com/pinage404/git-gamble/-/jobs/3351015661

Related to https://github.com/yusdacra/nix-cargo-integration/issues/53

How to use `installShellFiles` with the new API ?

**Comment by 90-008** [comment]
From your `update_nix_dependencies` branch, your `pkgConfig` should look like this:
```nix
      pkgConfig = common: {
        "git-gamble" = {
          build = true;
          app = true;

          overrides = {
            add-build-reqs.overrideAttrs = old: {
              nativeBuildInputs =
              	(old.nativeBuildInputs or [])
              	++ (with common.pkgs; [
                	installShellFiles
                	bash
                	git
              	]);
              preCheck = ''
                patchShebangs tests/editor/fake_editor.sh
              '';
              postInstall = ''
                echo $PATH
                type installShellCompletion
                installShellCompletion --fish target/$cargoBuildProfile/build/git-gamble-*/out/git-gamble.fish
                installShellCompletion --bash target/$cargoBuildProfile/build/git-gamble-*/out/git-gamble.bash
                installShellCompletion --zsh target/$cargoBuildProfile/build/git-gamble-*/out/_git-gamble

                PATH="$PATH:target/release/bin/" sh ./script/usage.sh > git-gamble.1
                installManPage git-gamble.1
              '';
            };
          };
        };
      };
```
This should work I believe. To explain, `overrides` for a package get applied to that packages *main derivation*, aka when your crate(s) are actually being built, as opposed to the *deps derivation* which only builds the dependencies. Since we want to install shell completion, man pages etc. these should be done in the main derivation, not the deps derivation. And `overrides` is basically just an attrset of dream2nix overrides. You can name them anything you want (here I chose `add-build-reqs`), and then you can specify the override attributes. Here I used `overrideAttrs`, since it lets you access previous derivation attributes.

**Reply by pinage404** [reply]
it still failing because it can't find `git` in the `$PATH` while running tests https://gitlab.com/pinage404/git-gamble/-/jobs/3380009041

with Rust, there is three types of dependencies : `build-dependencies`, `dependencies` and `dev-dependencies`

git-gamble needs : 
* `installShellFiles` while installing (kindof `build-dependencies`)
* `git` and `bash` when running tests (kindof `dev-dependencies`)
* `git` at runtime (kindof `dependencies`)

i don't know how to specify this different types of dependencies with Nix

**Comment by 90-008** [comment]
i'm not sure if this is is still an issue or not but since overhauling the options it's most likely not (you'd have to add such "main package" stuff to `drvConfig` attribute for a crate / project

**Reply by pinage404** [reply]
I don't remember which issues i had

I [stopped using it](https://gitlab.com/pinage404/git-gamble/-/commit/fbf2a2b5af641c7ba1834ba266e8cff2b5ee7310), since a while

Now, i use [rust-overlay](https://gitlab.com/pinage404/git-gamble/-/blob/8257946207f63aedfb88e00268607171ecfba1c4/flake.nix#L18) and [crane](https://gitlab.com/pinage404/git-gamble/-/blob/8257946207f63aedfb88e00268607171ecfba1c4/flake.nix#L23)
